test/merkleblock.js

```diff
@@ -0,0 +1,192 @@
+        return new MerkleBlock();
+      }).should.throw('Unrecognized argument for MerkleBlock');
+    });
+  });
+
+  describe('#fromJSON', function() {
+
+    it('should set these known values', function() {
+      var block = MerkleBlock.fromJSON(blockJSON);
+      should.exist(block.header);
+      should.exist(block.numTransactions);
+      should.exist(block.hashes);
+      should.exist(block.flags);
+    });
+
+    it('should set these known values', function() {
+      var block = MerkleBlock(blockJSON);
+      should.exist(block.header);
+      should.exist(block.numTransactions);
+      should.exist(block.hashes);
+      should.exist(block.flags);
+    });
+
+    it('accepts an object as argument', function() {
+      var block = MerkleBlock(blockbuf);
+      MerkleBlock.fromJSON(block.toObject()).should.exist();
+    });
+
+  });
+
+  describe('#toJSON', function() {
+
+    it('should recover these known values', function() {
+      var block = MerkleBlock.fromJSON(blockJSON);
+      var b = JSON.parse(block.toJSON());
+      should.exist(block.header);
+      should.exist(block.numTransactions);
+      should.exist(block.hashes);
+      should.exist(block.flags);
+      should.exist(b.header);
+      should.exist(b.numTransactions);
+      should.exist(b.hashes);
+      should.exist(b.flags);
+    });
+
+  });
+
+  describe('#fromBuffer', function() {
+
+    it('should make a block from this known buffer', function() {
+      var block = MerkleBlock.fromBuffer(blockbuf);
+      block.toBuffer().toString('hex').should.equal(blockhex);
+    });
+
+  });
+
+  describe('#fromBufferReader', function() {
+
+    it('should make a block from this known buffer', function() {
+      var block = MerkleBlock.fromBufferReader(BufferReader(blockbuf));
+      block.toBuffer().toString('hex').should.equal(blockhex);
+    });
+
+  });
+
+  describe('#toBuffer', function() {
+
+    it('should recover a block from this known buffer', function() {
+      var block = MerkleBlock.fromBuffer(blockbuf);
+      block.toBuffer().toString('hex').should.equal(blockhex);
+    });
+
+  });
+
+  describe('#toBufferWriter', function() {
+
+    it('should recover a block from this known buffer', function() {
+      var block = MerkleBlock.fromBuffer(blockbuf);
+      block.toBufferWriter().concat().toString('hex').should.equal(blockhex);
+    });
+
+    it('doesn\'t create a bufferWriter if one provided', function() {
+      var writer = new BufferWriter();
+      var block = MerkleBlock.fromBuffer(blockbuf);
+      block.toBufferWriter(writer).should.equal(writer);
+    });
+
+  });
+
+
+  describe('#validMerkleTree', function() {
+
+    it('should validate good merkleblocks', function() {
+      data.JSON.forEach(function(json) {
+        var b = MerkleBlock(JSON.stringify(json));
+        b.validMerkleTree().should.equal(true);
+        b._validMerkleTree.should.equal(true);
+      });
+    });
+
+    it('should respect _validMerkleTrees', function() {
```

Comment:
The fact that this can return "true" when it could actually be "false" because of the cache value, seems a bit concerning.